File: JavaAlgorithm/src/main/java/algo/tzashinorpu/SpecialSubject/Array_List/PreDiff/corpFlightBookings_1109.java
package algo.tzashinorpu.SpecialSubject.Array_List.PreDiff;

public class corpFlightBookings_1109 {
    public int[] corpFlightBookings_1(int[][] bookings, int n) {
        int[] ans = new int[n];
        for (int[] booking : bookings) {
            int start = booking[0] - 1;
            int end = booking[1] - 1;
            for (int i = start; i <= end; i++) {
                ans[i] += booking[2];
            }
        }
        return ans;
    }

    public int[] corpFlightBookings_2(int[][] bookings, int n) {
        int[] ans = new int[n];
        for (int[] booking : bookings) {
            ans[booking[0] - 1] += booking[2];
            if (booking[1] < n) ans[booking[1]] -= booking[2];
        }
        for (int i = 1; i < n; i++) {
            ans[i] += ans[i - 1];
        }
        return ans;
    }

    public int[] corpFlightBookings_3(int[][] bookings, int n) {
        int[] ans = new int[n];
        for (int[] booking : bookings) {
            int start = booking[0] - 1;
            int end = booking[1] - 1;
            int val = booking[2];
            ans[start] += val;
            if (end + 1 < n) ans[end + 1] += -val;
        }
        for (int i = 1; i < n; i++) {
            ans[i] += ans[i - 1];
        }
        return ans;
    }
}

File: Clion/ZeroTrac/1500_1600/corpFlightBookings_1109.cpp
//
// Created by TzashiNorpu on 7/20/2022.
//
#include "header.h"

using namespace ZeroTrac;

vector<int> corpFlightBookings_1109::corpFlightBookings(vector<vector<int>> &bookings, int n) {
    vector<int> res(n, 0);
    for (auto booking: bookings) {
        int b = booking[0] - 1;
        int e = booking[1] - 1;
        int v = booking[2];
        res[b] += v;
        if (e + 1 < n) res[e + 1] -= v;
    }
    for (int i = 1; i < n; ++i) {
        res[i] += res[i - 1];
    }
    return res;
}